cnap/core/rtdb.py

```diff
@@ -0,0 +1,154 @@
+        """
+        Save a dict value for an object in a table
+        """
+
+    @abstractmethod
+    def get_table_object_dict(self, table: str, obj: str) -> dict:
+        """
+        Get a dict value for an object from a table
+        """
+
+    @abstractmethod
+    def get_all_table_objects_dict(self, table: str) -> dict:
+        """
+        Get all dict values from a table
+        """
+
+    @abstractmethod
+    def check_table_object_exist(self, table: str, obj: str) -> bool:
+        """
+        Check whether given object exists in given table
+        """
+
+    @abstractmethod
+    def del_table_object(self, table: str, obj: str) -> None:
+        """
+        Delete an object from a given table
+        """
+
+class RedisDB(RuntimeDatabaseBase):
+
+    """
+    Redis backend implementation for runtime database
+    """
+
+    def __init__(self):
+        self._conn = None
+
+    def connect(self, host: str = "127.0.0.1", port: int = 6379, db: int = 0) -> bool:
+        """
+        Connect to Redis server.
+        Args:
+            host: Redis server host hostname/IP
+            port: Redis server port
+            db: Database number in Redis server
+        Returns:
+            bool: Indicate whether connecting to the Redis server successfully
+        """
+        self._conn = redis.Redis(host=host, port=port, db=db)
+        try:
+            self._conn.ping()
+        except redis.exceptions.ConnectionError:
+            LOG.error("Failed to connect to Redis")
+            return False
+        return True
+
+    def save_table_object_dict(self, table: str, obj: str, d: dict) -> None:
+        """
+        Save a dict value for an object in a table.
+        Args:
+            table: The name of the table
+            obj: The name of the object
+            d: The value to be saved
+        Returns:
+            None
+        """
+        LOG.debug("[Redis] Save => table: %s, obj: %s", table, obj)
+        self._conn.hset(table, obj, json.dumps(d))
+
+    def get_table_object_dict(self, table: str, obj: str) -> dict:
+        """
+        Get a dict value for an object from a table.
+        Args:
+            table: The name of the table
+            obj: The name of the object
+        Returns:
+            dict: The value get by table name and object name
+        """
+        value = self._conn.hget(table, obj)
+        if value is None:
+            return {}
+        try:
+            return json.loads(value)
+        except json.JSONDecodeError:
+            LOG.error("Invalid JSON value for table: %s, obj: %s", table, obj)
+            return {}
+
+    def get_all_table_objects_dict(self, table: str) -> dict:
+        """
+        Get all dict values from a table.
+        Args:
+            table: The name of the table
+        Returns:
+            dict: All values get by the table name
+        """
+        result_bytes = self._conn.hgetall(table)
```

Comment:
Do you think it would be better to add the error handler for each redis execution? Try the hegt, hgetall, hexists, hdel and catch the necessary errors. Same for the following method.

Reply:
Yes. After check, this command itself won't raise errors. Here we only may meet the redis connection issue, table type error or conn not a valid redis instance error.

So what do you think, if it is necessary to handle some of the errors?

Reply:
> Yes. After check, this command itself won't raise errors. Here we only may meet the redis connection issue, table type error or conn not a valid redis instance error.
> 
> So what do you think, if it is necessary to handle some of the errors?

Yes, I think we should handle at least one common error, I'm going to choose one of them.